gen2-rgbd-projection/main.py

```diff
@@ -0,0 +1,330 @@
+        pixel_coords = pixel_coord_np(res["width"], res["height"]).astype(float)
+
+        # main stream loop
+        print("Begin streaming at resolution: {} x {}".format(
+            res["width"], res["height"]))
+        while True:
+            for i, queue in enumerate(queue_list):
+                name = queue.getName()
+                image = queue.get()
+                if streams[i] == "rgb":
+                    data, w, h = image.getData(), image.getWidth(), image.getHeight()
+                    yuv = np.array(data).reshape(
+                        (h * 3 // 2, w)).astype(np.uint8)
+                    # pcl_frames[i] = cv2.cvtColor(yuv, cv2.COLOR_YUV2BGR_NV12)
+                    pcl_frames[i] = cv2.cvtColor(yuv, cv2.COLOR_YUV2RGB_NV12)
+                    pcl_frames[i] = cv2.resize(
+                        pcl_frames[i], (res["width"], res["height"]), cv2.INTER_CUBIC)
+                else:
+                    depth_frame = np.array(image.getFrame())
+                    if align_depth:
+                        if not align_on_host:
+                            pcl_frames[i] = cv2.resize(
+                                depth_frame, (res["width"], res["height"]), cv2.INTER_NEAREST)
+                        else:
+                            # Approach 1: Use opencv's register depth method
+                            depth_scale = 10.0  # extrinsic is in cm, but depth in mm so convert
+                            depth_dilation = True
+                            scaled_depth_frame = depth_frame.astype(
+                                float) / depth_scale
+                            pcl_frames[i] = cv2.rgbd.registerDepth(
+                                rectified_right_intrinsic, rgb_intrinsic, rgb_distortion, right_to_rgb_extrinsic, scaled_depth_frame, (res["width"], res["height"]), depth_dilation)
+                            pcl_frames[i] = np.array(
+                                pcl_frames[i] * depth_scale).astype(np.uint16)  # convert back from cm to mm
+                            # Approach 2: Use slow but easy to interpert function
+                            # pcl_frames[i] = alignDepthToRGB(depth_frame, pixel_coords, inverse_rectified_right_intrinsic, rgb_intrinsic, right_to_rgb_extrinsic, res["width"], res["height"])
+                    else:
+                        pcl_frames[i] = depth_frame
+
+            if any([frame is None for frame in pcl_frames]) and align_depth:
+                print("Error: Need rgb AND depth frame to align!")
+                continue
+            elif pcl_frames[0] is None:
+                print("Waiting on depth image to visualize")
+                continue
+
+            if align_depth:
+                # Convert nonzero depth into red pixels in 3-channel image
+                depth_three_channel = np.zeros_like(pcl_frames[1])
+                depth_three_channel[:, :, 2] = (
+                    255 * pcl_frames[0].astype(float) / max_disparity).astype(np.uint8)
```

Comment:
is this for visualization ?

Reply:
Yes, purely for visualization. here I am turning the depth image into a 3-channel image (where the depth values > 0 are red pixels and all others are white) so it can be overlaid onto the rgb image. This allows the user to visually verify the rgb-d alignment (since it can be cumbersome to do by looking at the pointcloud).

Reply:
Also moving these utility parts of the code to functions in a separate file to de-clutter.